### PR TITLE
Support Animated Capes

### PR DESCRIPTION
**Pull request by james090500**
This supports animated capes using a sprite. [Example here](https://minecraftcapes.net/gallery/3070eec3afaaa9a3fd5cef2546701818d4c052f3a71b744e91747858d307cb86)

Will probably need a manual merge and also needs this to be merged before hand
https://github.com/bs-community/skinview-utils/pull/15

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,9 +1,9 @@
 skinview3d
 ========
 
-[![CI Status](https://img.shields.io/github/workflow/status/bs-community/skinview3d/CI?label=CI&logo=github&style=flat-square)](https://github.com/bs-community/skinview3d/actions?query=workflow:CI)
+[![CI Status](https://img.shields.io/github/workflow/status/james090500/skinview3d/CI?label=CI&logo=github&style=flat-square)](https://github.com/james090500/skinview3d/actions?query=workflow:CI)
```
Comment (Hacksore) [inline README.md:4]: This should be reverted back to `bs-community`

File: README.md [hunk 2]
```diff
@@ -1,9 +1,9 @@
...
 [![NPM Package](https://img.shields.io/npm/v/skinview3d.svg?style=flat-square)](https://www.npmjs.com/package/skinview3d)
-[![MIT License](https://img.shields.io/badge/license-MIT-yellowgreen.svg?style=flat-square)](https://github.com/bs-community/skinview3d/blob/master/LICENSE)
+[![MIT License](https://img.shields.io/badge/license-MIT-yellowgreen.svg?style=flat-square)](https://github.com/james090500/skinview3d/blob/master/LICENSE)
```
Comment (Hacksore) [inline README.md:6]: This should be reverted back to `bs-community`

File: README.md [hunk 3]
```diff
@@ -92,7 +92,7 @@ will completely disable shadows.
 
 ## Ears
 skinview3d supports two types of ear texture:
-* `standalone`: 14x7 image that contains the ear ([example](https://github.com/bs-community/skinview3d/blob/master/examples/img/ears.png))
+* `standalone`: 14x7 image that contains the ear ([example](https://github.com/james090500/skinview3d/blob/master/examples/img/ears.png))
```
Comment (Hacksore) [inline README.md:95]: This should be reverted back to `bs-community`

File: package.json [hunk 4]
```diff
@@ -1,5 +1,5 @@
 {
-	"name": "skinview3d",
+	"name": "@james090500/skinview3d",
```
Comment (Hacksore) [inline package.json:2]: This should be reverted back to `skinview3d`

File: package.json [hunk 5]
```diff
@@ -30,32 +30,32 @@
 	],
 	"license": "MIT",
 	"bugs": {
-		"url": "https://github.com/bs-community/skinview3d/issues"
+		"url": "https://github.com/james090500/skinview3d/issues"
```
Comment (Hacksore) [inline package.json:33]: This should be reverted back to `bs-community`

File: package.json [hunk 6]
```diff
@@ -30,32 +30,32 @@
...
 	},
-	"homepage": "https://github.com/bs-community/skinview3d",
+	"homepage": "https://github.com/james090500/skinview3d",
```
Comment (Hacksore) [inline package.json:35]: This should be reverted back to `bs-community`

File: rollup.config.js [hunk 7]
```diff
@@ -9,7 +9,7 @@ export default {
         file: "bundles/skinview3d.bundle.js",
         format: "umd",
         name: "skinview3d",
-        banner: "/* @preserve skinview3d / MIT License / https://github.com/bs-community/skinview3d */",
+        banner: "/* @preserve skinview3d / MIT License / https://github.com/james090500/skinview3d */",
```
Comment (Hacksore) [inline rollup.config.js:12]: This should be reverted back to `bs-community`

File: src/skinview-overrides.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+		// 64x32
+		return image.width / 64;
+	} else if (image.width * 17 === image.height * 22) {
+		// 22x17
+		return image.width / 22;
+	} else if (image.width * 11 === image.height * 23) {
+		// 46x22
+		return image.width / 46;
+    } else if ((image.height % (image.width / 2)) == 0) {
+		//Animated cape HD with raito 64x32
```
Comment (Hacksore) [inline src/skinview-overrides.ts:15]: nit: a space before starting comment text

File: src/viewer.ts [hunk 9]
```diff
@@ -9,6 +9,9 @@ import { PlayerAnimation } from "./animation.js";
 import { BackEquipment, PlayerObject } from "./model.js";
 import { NameTagObject } from "./nametag.js";
 
+//james090500
+import { loadCapeToCanvas } from "./skinview-overrides.js";
```
Comment (Hacksore) [inline src/viewer.ts:13]: This probably should be added to `skinview-utils` right?

File: src/viewer.ts [hunk 10]
```diff
@@ -346,7 +349,7 @@ export class SkinViewer {
 		let renderTarget;
 		if (this.renderer.capabilities.isWebGL2) {
 			// Use float precision depth if possible
-			// see https://github.com/bs-community/skinview3d/issues/111
+			// see https://github.com/james090500/skinview3d/issues/111
```
Comment (Hacksore) [inline src/viewer.ts:352]: This should be reverted back to `bs-community`

File: src/viewer.ts [hunk 11]
```diff
@@ -524,6 +527,29 @@ export class SkinViewer {
 		}
 	}
 
+	//james090500
+	private frameDelay = 100;
```
Comment (Hacksore) [inline src/viewer.ts:531]: Should this be controllable by the consumer?

**Review by Hacksore — CHANGES_REQUESTED** [review]

File: src/skinview-overrides.ts (hunk 8 above)
Comment (james090500) [reply src/skinview-overrides.ts:15]: Moved to https://github.com/bs-community/skinview-utils/pull/15

File: src/viewer.ts (hunk 9 above)
Comment (james090500) [reply src/viewer.ts:13]: Moved to https://github.com/bs-community/skinview-utils/pull/15

File: src/viewer.ts (hunk 11 above)
Comment (james090500) [reply src/viewer.ts]: I believe this has been added

File: rollup.config.js [hunk 12]
```diff
@@ -1,6 +1,6 @@
 import resolve from "@rollup/plugin-node-resolve";
 import typescript from '@rollup/plugin-typescript';
-import { terser } from "rollup-plugin-terser";
+import terser from '@rollup/plugin-terser';
```
Comment (Hacksore) [inline rollup.config.js:3]: This should be double quotes to match code style

**Review by Hacksore — CHANGES_REQUESTED** [review]
Just minor format issue and I can give a 👍 once you fix that.

**Review by Hacksore — APPROVED** [review]
LGTM, needs review from @yushijinhun

**Comment by james090500** [comment]
Any update on this?

File: src/viewer.ts [hunk 13]
```diff
@@ -524,6 +535,32 @@ export class SkinViewer {
...
+	private capeOptions: CapeLoadOptions = {}
+	private totalFrames = 1;
+	private currentFrame = 1;
+	animatedCape(): void {
+		const currentTime: number = Date.now();
+		if(this.lastFrameTime + +this._frameDelay > currentTime) return;
+		this.lastFrameTime = currentTime
+
+		if(this.capeSource != null && this.totalFrames > 1) {
+			this.loadCape(this.capeSource, this.capeOptions)
```
Comment (yushijinhun) [inline src/viewer.ts:549]: Calling `loadCape()` on each frame update would have a huge performance impact. I think it's better to load each frame into a individual canvas and just switch the canvas periodically.

**Comment by Hacksore** [comment]
@james090500 has implemented this in this repo ❤️. Going to close this for now as it's a bit hard to get merged in.

https://github.com/minecraftcapes/minecraft-skin-viewer

**Pull request closed without merging**